core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -0,0 +1,561 @@
+
+    private final OakComparator<K> comparator;
+
+    // use reference codec to encode Full Hash Number integer (first) with its updates counter (second)
+    private final ReferenceCodec hashCodec = new ReferenceCodec(
+        FULL_HASH_BITS, // bits# to represent full hash number as integer, bits# to represent update
+        ReferenceCodec.INVALID_BIT_SIZE); // counter are calculated upon previous parameters (also int)
+
+    /*----------------- Constructor -------------------*/
+
+    /**
+     * Create a new EntryHashSet
+     * @param vMM   for values off-heap allocations and releases
+     * @param kMM off-heap allocations and releases for keys
+     * @param entriesCapacity how many entries should this EntryOrderedSet keep at maximum
+     * @param keySerializer   used to serialize the key when written to off-heap
+     */
+    EntryHashSet(MemoryManager vMM, MemoryManager kMM, int entriesCapacity, OakSerializer<K> keySerializer,
+        OakSerializer<V> valueSerializer, OakComparator<K> comparator) {
+        // We add additional field for the head (dummy) node
+        super(vMM, kMM, ADDITIONAL_FIELDS, entriesCapacity, keySerializer, valueSerializer);
+        this.comparator = comparator;
+    }
+
+
+
+    /********************************************************************************************/
+    /*---------- Methods for managing full hash entry indexes (package visibility) -------------*/
+
+    /**
+     * getFullHashNumber returns the full hash entry index (the number only!) of the entry
+     * given by entry index "ei". The method serves internal and external EntryHashSet users.
+     */
+    private int getFullHashNumber(int ei) {
+        assert isIndexInBound(ei);
+        long fullHashNumberField = getEntryFieldLong(ei, HASH_FIELD_OFFSET);
+        // extract the hash number (first) from the updates counter (second)
+        return hashCodec.getFirst(fullHashNumberField);
+    }
+
+    /**
+     * getFullHashEntryField returns the full hash entry index (the entire field including counter!)
+     * of the entry given by entry index "ei". The method serves internal and external EntryHashSet users.
+     */
+    private long getFullHashEntryField(int ei) {
+        assert isIndexInBound(ei);
+        return getEntryFieldLong(ei, HASH_FIELD_OFFSET);
+    }
+
+    /**
+     * isFullHashIndexValid checks the index itself disregarding the update counter
+     */
+    private boolean isFullHashIndexValid(int ei) {
+        return (getFullHashNumber(ei) != INVALID_FULL_HASH);
+    }
+
+    /**
+     * setFullHashEntryIndex sets the full hash entry index (of the entry given by entry index "ei")
+     * to be the "fullHash". To be used while rebalance, during normal path only CAS is used.
+     */
+    private void setFullHashEntryIndex(int ei, long fullHash) {
+        setEntryFieldLong(ei, HASH_FIELD_OFFSET, fullHash);
+    }
```

Comment:
I understand this sets the entry's hash, but the name suggests that it sets the entry index of the hash, which doesn't make sense.
Did you mean `setHash(int ei, long hash)`?
Just as we named the other methods: `setKeyReference(int ei, long value)` etc.

Anyway, it is never used.

Reply:
The name will be clarified when we agree about the issues explained above. Skipping.